### PR TITLE
Add tags to session

### PR DESCRIPTION
**Pull request by rongrong**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 1]
```diff
@@ -222,6 +228,18 @@ public Session createSession(
         return sessionProperties;
     }
 
+    private Set<String> parseTags(HttpServletRequest servletRequest)
+    {
+        String tags = servletRequest.getHeader(PRESTO_TAGS);
+        if (tags != null) {
+            Splitter splitter = Splitter.on(',').trimResults().omitEmptyStrings();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: You can simplify this code by doing nullToEmpty() on "tags". Then you don't need an if statement.

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 2]
```diff
@@ -352,6 +364,7 @@ public String toString()
                 .add("remoteUserAddress", remoteUserAddress.orElse(null))
                 .add("userAgent", userAgent.orElse(null))
                 .add("clientInfo", clientInfo.orElse(null))
+                .add("tags", Joiner.on(',').join(tags))
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: Use the default formatting for sets.

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 3]
```diff
@@ -96,6 +100,7 @@ public Session(
         this.remoteUserAddress = requireNonNull(remoteUserAddress, "remoteUserAddress is null");
         this.userAgent = requireNonNull(userAgent, "userAgent is null");
         this.clientInfo = requireNonNull(clientInfo, "clientInfo is null");
+        this.clientTags = requireNonNull(clientTags, "clientTags is null");
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: safer to create an immutable copy (similar to what we do with `systemProperties` here).

File: presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java [hunk 4]
```diff
@@ -45,6 +46,7 @@
     private final Optional<String> remoteUserAddress;
     private final Optional<String> userAgent;
     private final Optional<String> clientInfo;
+    private final Set<String> tags;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java]: we call this `clientTags` in `Session` we can use the same name here to be consistent.

File: presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java [hunk 5]
```diff
@@ -83,6 +86,7 @@ public SessionRepresentation(
         this.remoteUserAddress = requireNonNull(remoteUserAddress, "remoteUserAddress is null");
         this.userAgent = requireNonNull(userAgent, "userAgent is null");
         this.clientInfo = requireNonNull(clientInfo, "clientInfo is null");
+        this.tags = requireNonNull(tags, "tags is null");
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/SessionRepresentation.java]: ditto about keeping an  immutable copy here.

File: presto-main/src/test/java/com/facebook/presto/server/TestHttpRequestSessionFactory.java [hunk 6]
```diff
@@ -58,6 +60,7 @@ public void testCreateSession()
                         .put(PRESTO_LANGUAGE, "zh-TW")
                         .put(PRESTO_TIME_ZONE, "Asia/Taipei")
                         .put(PRESTO_CLIENT_INFO, "client-info")
+                        .put(PRESTO_TAGS, "tag1,tag2,tag3")
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/server/TestHttpRequestSessionFactory.java]: please add tests for empty and null tags.

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java [hunk 7]
```diff
@@ -222,6 +228,12 @@ public Session createSession(
         return sessionProperties;
     }
 
+    private Set<String> parseTags(HttpServletRequest servletRequest)
+    {
+        Splitter splitter = Splitter.on(',').trimResults().omitEmptyStrings();
+        return ImmutableSet.copyOf(splitter.splitToList(nullToEmpty(servletRequest.getHeader(PRESTO_TAGS))));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: do you think we should have some cap on the number of tags that we allow to be set on http requests?

Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionFactory.java]: It's sort of reasonable, but I don't know what's a reasonable number. 10? 50? 100? (10 seems to be to small though)

**Comment by rongrong** [comment]
@electrum @nezihyigitbasi Could you guys take a look? Thanks!

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 8]
```diff
@@ -225,6 +236,12 @@ public boolean supportClientTransaction()
         return sessionProperties;
     }
 
+    private Set<String> parseClientTags(HttpServletRequest servletRequest)
+    {
+        Splitter splitter = Splitter.on(',').trimResults().omitEmptyStrings();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: This disallows tags with commas. Is that ok? Do we want any other restrictions?

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java [hunk 9]
```diff
@@ -225,6 +236,12 @@ public boolean supportClientTransaction()
...
+        return ImmutableSet.copyOf(splitter.splitToList(nullToEmpty(servletRequest.getHeader(PRESTO_CLIENT_TAGS))));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: Nit: use `splitter.split()`, so we don't need to construct a collection twice

File: presto-main/src/test/java/com/facebook/presto/server/TestQuerySessionSupplier.java [hunk 10]
```diff
@@ -57,6 +59,7 @@ public void testCreateSession()
                         .put(PRESTO_LANGUAGE, "zh-TW")
                         .put(PRESTO_TIME_ZONE, "Asia/Taipei")
                         .put(PRESTO_CLIENT_INFO, "client-info")
+                        .put(PRESTO_CLIENT_TAGS, "tag1,tag2,tag3")
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/server/TestQuerySessionSupplier.java]: Add a duplicate in the list

File: presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java (hunk 8 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/server/HttpRequestSessionContext.java]: I think it's acceptable to disallow commas in tags.

**Pull request closed without merging**